File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/fused_moe_e2e_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/fused_moe_e2e_emit.c -- C-side emitter for the fused_moe_e2e
 * (end-to-end fused-MoE forward orchestrator) parity harness.
 *
 * Selects one of the sampled FusedMoeForwardSpec configs by argv[1] (the config
 * index), materialises the spec via rocke_fmoe_forward_spec_default() + the per-
 * config shape fields, then lowers each lowerable pipeline stage to AMDGPU LLVM
 * IR text via rocke_fused_moe_forward_lower_to_llvm (arch gfx950, flavor AUTO) and
 * prints them concatenated to stdout, each prefixed with a stage banner, so the
 * output can be byte-compared with the Python emitter fused_moe_e2e_emit.py.
 *
 * The orchestrator emits NO single monolithic kernel; rocke_fused_moe_forward_
 * lower_to_llvm runs __init__'s arch resolve + tile-swap policy internally (via
 * the build ctx) and delegates to the spec-selected sub-kernel builder. The
 * three lowerable stages are ROUTER (topk_softmax) and GATE_UP_GEMM / DOWN_GEMM
 * (both the batched-GEMM builder shape). The remaining stages (sort hist/scan/
 * scatter, gather, silu_mul, topk_reduce) are TODO(port) NOTIMPL and are
 * therefore excluded here, matching the Python emitter.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "rocke/instance_batched_gemm.h"
#include "rocke/instance_fused_moe_e2e.h"
#include "rocke/instance_fused_moe_e2e_internal.h"
#include "rocke/instance_topk_softmax.h"
#include "rocke/ir.h"
#include "rocke/ir_serialize.h"
#include "rocke/lower_llvm.h"
#include "rocke/verify.h"

/* Populate the FusedMoeForwardSpec for config `idx`. Returns 0, or -1 on an
 * unknown index. Every non-shape field stays at the dataclass default
 * (rocke_fmoe_forward_spec_default), so the tile-swap policy and the static gate
 * see exactly the Python defaults; only the enumerated shape + dtype differ. */
static int make_spec(int idx, rocke_fmoe_forward_spec_t* spec)
{
    *spec = rocke_fmoe_forward_spec_default();
    spec->arch = "gfx950";
    switch(idx)
    {
    case 0: /* tokens=1   E8  K2 H4096 I7168 f16  */
        spec->tokens = 1;
        spec->experts = 8;
        spec->topk = 2;
        spec->hidden = 4096;
        spec->intermediate = 7168;
        spec->dtype = "f16";
        break;
    case 1: /* tokens=8   E8  K2 H4096 I7168 f16  */
        spec->tokens = 8;
        spec->experts = 8;
        spec->topk = 2;
        spec->hidden = 4096;
        spec->intermediate = 7168;
        spec->dtype = "f16";
        break;
    case 2: /* tokens=32  E8  K2 H4096 I7168 f16  */
        spec->tokens = 32;
        spec->experts = 8;
        spec->topk = 2;
        spec->hidden = 4096;
        spec->intermediate = 7168;
        spec->dtype = "f16";
        break;
    case 3: /* tokens=128 E8  K2 H4096 I7168 f16  */
        spec->tokens = 128;
        spec->experts = 8;
        spec->topk = 2;
        spec->hidden = 4096;
        spec->intermediate = 7168;
        spec->dtype = "f16";
        break;
    case 4: /* tokens=1   E8  K2 H4096 I7168 bf16 */
        spec->tokens = 1;
        spec->experts = 8;
        spec->topk = 2;
        spec->hidden = 4096;
        spec->intermediate = 7168;
        spec->dtype = "bf16";
        break;
    case 5: /* tokens=128 E32 K5 H8192 I8192 f16  */
        spec->tokens = 128;
        spec->experts = 32;
        spec->topk = 5;
        spec->hidden = 8192;
        spec->intermediate = 8192;
        spec->dtype = "f16";
        break;
    default:
        return -1;
    }
    return 0;
}

/* Lowerable stages, in the Python emitter's order. GATE_UP_GEMM and DOWN_GEMM
 * both resolve to the batched-GEMM builder shape inside
 * rocke_fused_moe_forward_lower_to_llvm. */
static const struct
{
    const char* banner;
    rocke_fmoe_stage_t stage;
} STAGES[] = {
    {"ROUTER", ROCKE_FMOE_STAGE_ROUTER},
    {"GATE_UP_GEMM", ROCKE_FMOE_STAGE_GATE_UP_GEMM},
    {"DOWN_GEMM", ROCKE_FMOE_STAGE_DOWN_GEMM},
};

/* Build a kernel for one stage using the tile-policy-adjusted spec.
 * The caller must rocke_ir_builder_free(&b) when done. Returns NULL on failure. */
static rocke_kernel_def_t* build_stage_kernel(rocke_ir_builder_t* b,
                                              const rocke_fmoe_forward_spec_t* spec,
                                              rocke_fmoe_stage_t stage,
                                              const char* banner)
{
    /* Apply tile-swap policy via the build ctx (same as lower_to_llvm). */
    rocke_fmoe_build_ctx_t* ctx = (rocke_fmoe_build_ctx_t*)calloc(1, sizeof(*ctx));
    if(!ctx)
        return NULL;
    rocke_status_t st = rocke_fmoe_build_ctx_init(ctx, spec, "gfx950");
    if(st != ROCKE_OK)
    {
        rocke_fmoe_build_ctx_destroy(ctx);
        free(ctx);
        return NULL;
    }
    rocke_fmoe_forward_spec_t adj = ctx->spec;
    const char* arch = ctx->arch;
    rocke_kernel_def_t* kernel = NULL;

    if(stage == ROCKE_FMOE_STAGE_ROUTER)
    {
        rocke_topk_softmax_spec_t s = rocke_fmoe_forward_spec_to_topk_softmax_spec(&adj);
        kernel = rocke_build_topk_softmax_new(b, &s, arch);
    }
    else if(stage == ROCKE_FMOE_STAGE_GATE_UP_GEMM || stage == ROCKE_FMOE_STAGE_DOWN_GEMM)
    {
        char name_buf[256];
        rocke_batched_gemm_spec_t s;
        st = rocke_fmoe_forward_spec_to_batched_gemm_spec(&adj, name_buf, sizeof(name_buf), &s);
        if(st == ROCKE_OK)
        {
            kernel = rocke_build_batched_gemm_new(b, &s, arch);
        }
    }
    else
    {
        fprintf(stderr, "build_stage_kernel: stage %s not wired for ir/verify\n", banner);
    }
    rocke_fmoe_build_ctx_destroy(ctx);
    free(ctx);
    return kernel;
}

int main(int argc, char** argv)
{
    if(argc < 2)
    {
        fprintf(stderr, "usage: %s <config_index 0..5>\n", argv[0]);
        return 2;
    }
    int idx = atoi(argv[1]);
    const char* mode = (argc > 2) ? argv[2] : "ll";

    rocke_fmoe_forward_spec_t spec;
    if(make_spec(idx, &spec) != 0)
    {
        fprintf(stderr, "unknown config index %d\n", idx);
        return 2;
    }

    const size_t nstages = sizeof(STAGES) / sizeof(STAGES[0]);

    if(strcmp(mode, "ll") == 0)
    {
        for(size_t i = 0; i < nstages; i++)
        {
            char* llvm_text = NULL;
            char err[ROCKE_ERR_MSG_CAP];
            err[0] = 0;
            rocke_status_t st = rocke_fused_moe_forward_lower_to_llvm(&spec,
                                                                      "gfx950",
                                                                      STAGES[i].stage,
                                                                      ROCKE_LLVM_FLAVOR_AUTO,
                                                                      &llvm_text,
                                                                      err,
                                                                      sizeof err);
            if(st != ROCKE_OK || !llvm_text)
            {
                fprintf(stderr,
                        "lower failed (config %d stage %s): status=%d err=%s\n",
                        idx,
                        STAGES[i].banner,
                        (int)st,
                        err);
                free(llvm_text);
                return 1;
            }
            printf("; === fused_moe_e2e stage: %s ===\n", STAGES[i].banner);
            fputs(llvm_text, stdout);
            size_t n = strlen(llvm_text);
            if(n == 0 || llvm_text[n - 1] != '\n')
            {
                fputc('\n', stdout);
            }
            free(llvm_text);
        }
        return 0;
    }

    /* ir / verify: build each stage kernel and serialize/verify */
    for(size_t i = 0; i < nstages; i++)
    {
        rocke_ir_builder_t b;
        rocke_kernel_def_t* kernel
            = build_stage_kernel(&b, &spec, STAGES[i].stage, STAGES[i].banner);
        if(!kernel)
        {
            fprintf(stderr, "build failed (config %d stage %s)\n", idx, STAGES[i].banner);
            rocke_ir_builder_free(&b);
            return 1;
        }
        printf("; === fused_moe_e2e stage: %s ===\n", STAGES[i].banner);
        if(strcmp(mode, "ir") == 0)
        {
            char* t = NULL;
            rocke_status_t st = rocke_ir_serialize(kernel, &t);
            if(st != ROCKE_OK || !t)
            {
                fprintf(stderr, "serialize failed: status=%d\n", (int)st);
                rocke_ir_builder_free(&b);
                return 1;
            }
            fputs(t, stdout);
            size_t n = strlen(t);
            if(n == 0 || t[n - 1] != '\n')
                fputc('\n', stdout);
            free(t);
        }
        else if(strcmp(mode, "verify") == 0)
        {
            rocke_diag_t* d = NULL;
            size_t n = 0;
            rocke_verify(kernel, &d, &n);
            for(size_t j = 0; j < n; j++)
            {
                char* s = rocke_diag_to_string(&d[j]);
                if(s)
                {
                    puts(s);
                    free(s);
                }
            }
            rocke_diags_free(d, n);
        }
        else
        {
            fprintf(stderr, "unknown mode %s\n", mode);
            rocke_ir_builder_free(&b);
            return 2;
        }
        rocke_ir_builder_free(&b);
    }
    return 0;
}

File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/fused_moe_e2e_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/fused_moe_e2e_emit.py -- Python reference emitter for the
# fused_moe_e2e (end-to-end fused-MoE forward orchestrator) parity harness.
#
# Selects one of the sampled FusedMoeForwardSpec configs by argv[1] (the config
# index), constructs FusedMoeForward(spec) so the __init__ shape-aware tile-swap
# policy + arch resolution runs exactly as the live orchestrator would, then
# lowers each lowerable pipeline stage to AMDGPU LLVM IR via lower_kernel_to_llvm
# and prints them concatenated to stdout, each prefixed with a stage banner, so
# the output can be byte-compared with the C emitter fused_moe_e2e_emit.c.
#
# The orchestrator emits NO single monolithic kernel; it composes pre-existing
# sub-kernel builders selected by the (tile-policy-adjusted) spec. The C port's
# rocke_fused_moe_forward_lower_to_llvm currently wires three lowerable stages --
# ROUTER (topk_softmax), GATE_UP_GEMM and DOWN_GEMM (both the batched-GEMM
# builder shape) -- so this emitter lowers exactly those three, in the same
# order, to keep the byte comparison apples-to-apples. (The sort / gather /
# silu_mul / topk_reduce stages return NOTIMPL on the C side and are therefore
# excluded from the diff here.)
import sys

try:
    from rocke.core.lower_llvm import _lower_kernel_to_llvm_python as _native_lower
except ImportError:  # pragma: no cover - older reference tree
    from rocke import lower_kernel_to_llvm as _native_lower
from rocke.instances.common.fused_moe_e2e import (
    FusedMoeForwardSpec,
    FusedMoeForward,
)
from rocke.instances.common.topk_softmax import build_topk_softmax
from rocke.instances.common.batched_gemm import build_batched_gemm
from rocke.core.ir_serialize import serialize
from rocke.core.verify import verify


# The six sampled configs (must match the C emitter's make_spec()).
_CONFIGS = (
    dict(tokens=1, experts=8, topk=2, hidden=4096, intermediate=7168, dtype="f16"),
    dict(tokens=8, experts=8, topk=2, hidden=4096, intermediate=7168, dtype="f16"),
    dict(tokens=32, experts=8, topk=2, hidden=4096, intermediate=7168, dtype="f16"),
    dict(tokens=128, experts=8, topk=2, hidden=4096, intermediate=7168, dtype="f16"),
    dict(tokens=1, experts=8, topk=2, hidden=4096, intermediate=7168, dtype="bf16"),
    dict(tokens=128, experts=32, topk=5, hidden=8192, intermediate=8192, dtype="f16"),
)


def _spec(idx: int) -> FusedMoeForwardSpec:
    if idx < 0 or idx >= len(_CONFIGS):
        raise SystemExit(f"unknown config index {idx}")
    cfg = _CONFIGS[idx]
    return FusedMoeForwardSpec(arch="gfx950", **cfg)


# Lowerable stages, in the C emitter's order. Each entry is (banner, kernel_builder)
# where the kernel_builder takes the (tile-policy-adjusted) FusedMoeForward instance
# and returns the kernel object (not text).
def _build_router_kernel(fwd: FusedMoeForward):
    spec = fwd.spec.to_topk_softmax_spec()
    return build_topk_softmax(spec)


def _build_batched_gemm_kernel(fwd: FusedMoeForward):
    # GATE_UP_GEMM and DOWN_GEMM both lower via the batched-GEMM builder shape
    # (the orchestrator parameterises it per-call; the representative .ll for
    # golden diffing is the batched_gemm builder), matching the C port.
    spec = fwd.spec.to_batched_gemm_spec()
    return build_batched_gemm(spec)


_STAGES = (
    ("ROUTER", _build_router_kernel),
    ("GATE_UP_GEMM", _build_batched_gemm_kernel),
    ("DOWN_GEMM", _build_batched_gemm_kernel),
)


def main() -> int:
    if len(sys.argv) < 2:
        sys.stderr.write("usage: fused_moe_e2e_emit.py <config_index 0..5>\n")
        return 2
    idx = int(sys.argv[1])
    mode = sys.argv[2] if len(sys.argv) > 2 else "ll"
    spec = _spec(idx)
    # Construct the orchestrator so __init__ runs the arch resolve + shape-aware
    # tile-swap policy; the stages below read the post-swap fwd.spec.
    fwd = FusedMoeForward(spec)
    out = []
    for banner, build_kernel in _STAGES:
        kernel = build_kernel(fwd)
        out.append(f"; === fused_moe_e2e stage: {banner} ===\n")
        if mode == "ll":
            text = _native_lower(kernel, arch="gfx950")
            out.append(text)
            if not text.endswith("\n"):
                out.append("\n")
        elif mode == "ir":
            text = serialize(kernel)
            out.append(text)
            if not text.endswith("\n"):
                out.append("\n")
        elif mode == "verify":
            diags = "".join(str(d) + "\n" for d in verify(kernel))
            out.append(diags)
        else:
            sys.stderr.write(f"unknown mode {mode}\n")
            return 2
    sys.stdout.write("".join(out))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
